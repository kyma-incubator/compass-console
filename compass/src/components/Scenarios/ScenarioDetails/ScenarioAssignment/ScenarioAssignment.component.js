import React from 'react';
import PropTypes from 'prop-types';

import { GenericList } from 'react-shared';
import { useMutation } from '@apollo/react-hooks';

import deleteScenarioAssignmentHandler from '../shared/deleteScenarioAssignmentHandler';
import { SEND_NOTIFICATION } from '../../../../gql';

ScenarioAssignment.propTypes = {
  scenarioName: PropTypes.string.isRequired,
  getRuntimesForScenario: PropTypes.object.isRequired,
  getScenarioAssignment: PropTypes.object.isRequired,
  deleteScenarioAssignment: PropTypes.func.isRequired,
};

export default function ScenarioAssignment({
  scenarioName,
  getRuntimesForScenario,
  getScenarioAssignment,
  deleteScenarioAssignment,
}) {
  const [sendNotification] = useMutation(SEND_NOTIFICATION);
  const NOT_FOUND_MSG = 'NotFound';

  let hasScenarioAssignment = true;
  if (getScenarioAssignment.loading) {
    return <p>Loading...</p>;
  }
  if (getScenarioAssignment.error) {
    let errorType =
      getScenarioAssignment.error.graphQLErrors[0].extensions.error;
    if (!errorType.includes(NOT_FOUND_MSG)) {
      return `Error! ${getScenarioAssignment.error.message}`;
    }

    hasScenarioAssignment = false;
  }

  const showSuccessNotification = scenarioAssignmentName => {
    sendNotification({
      variables: {
        content: `Removed automatic scenario assignment from "${scenarioName}".`,
        title: `Successfully removed`,
        color: '#359c46',
        icon: 'accept',
        instanceName: scenarioName,
      },
    });
  };

  const actions = [
    {
      name: 'Delete',
      handler: async scenarioAssignment => {
        await deleteScenarioAssignmentHandler(
          deleteScenarioAssignment,
          scenarioName,
          async () => {
            showSuccessNotification(scenarioAssignment.selector.key);
            await getScenarioAssignment.refetch();
            await getRuntimesForScenario.refetch();
          },
        );
      },
    },
  ];

  const scenarioAssignments = [];
  if (hasScenarioAssignment) {
    scenarioAssignments[0] =
      getScenarioAssignment.automaticScenarioAssignmentForScenario;
  }

  return (
    <GenericList
      title="Automatic Scenario Assignment"
      notFoundMessage="No Automatic Scenario Assignment for this Scenario"
      entries={scenarioAssignments}
      headerRenderer={() => ['Key', 'Value']}
      actions={actions}
      rowRenderer={asa => [asa.selector.key, asa.selector.value]}
      showSearchField={false}
    />
  );
}
