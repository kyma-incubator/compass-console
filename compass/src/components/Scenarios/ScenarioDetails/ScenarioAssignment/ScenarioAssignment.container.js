import { graphql } from 'react-apollo';
import { compose } from 'recompose';
import { fromRenderProps } from 'recompose';
import {
  GET_ASSIGNMENT_FOR_SCENARIO,
  DELETE_ASSIGNMENT_FOR_SCENARIO,
  GET_RUNTIMES_FOR_SCENARIO,
  createEqualityQuery,
} from '../../gql';
import { SEND_NOTIFICATION } from '../../../../gql';

import ScenarioAssignment from './ScenarioAssignment.component';
import ScenarioNameContext from './../ScenarioNameContext';

export default compose(
  fromRenderProps(ScenarioNameContext.Consumer, scenarioName => ({
    scenarioName,
  })),
  graphql(DELETE_ASSIGNMENT_FOR_SCENARIO, {
    props: ({ mutate }) => ({
      deleteScenarioAssignment: async scenarioName =>
        await mutate({
          variables: { scenarioName },
        }),
    }),
  }),
  graphql(GET_ASSIGNMENT_FOR_SCENARIO, {
    name: 'getScenarioAssignment',
    options: ({ scenarioName }) => {
      return {
        errorPolicy: 'all',
        variables: {
          scenarioName: scenarioName,
        },
      };
    },
  }),
  graphql(GET_RUNTIMES_FOR_SCENARIO, {
    name: 'getRuntimesForScenario',
    options: ({ scenarioName }) => {
      const filter = {
        key: 'scenarios',
        query: createEqualityQuery(scenarioName),
      };
      return {
        variables: {
          filter: [filter],
        },
      };
    },
  }),
  graphql(SEND_NOTIFICATION, {
    name: 'sendNotification',
  }),
)(ScenarioAssignment);
