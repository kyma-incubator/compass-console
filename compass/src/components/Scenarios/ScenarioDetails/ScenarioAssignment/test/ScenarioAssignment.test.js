import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ScenarioAssignment from '../ScenarioAssignment.component';
import { runtimeResponseMock } from './mock';
import { assignmentResponseMock } from './mock';
jest.mock('react-shared', () => ({
  GenericList: function GenericListMocked(props) {
    return 'generic-list-mocked-content';
  },
}));

describe('ScenarioAssignment', () => {
  it('Renders with minimal props', () => {
    const component = shallow(
      <ScenarioAssignment
        scenarioName="test-scenario"
        getRuntimesForScenario={runtimeResponseMock}
        getScenarioAssignment={assignmentResponseMock}
        deleteScenarioAssignment={() => {}}
        sendNotification={() => {}}
      />,
    );

    expect(toJson(component)).toMatchSnapshot();
  });
});
