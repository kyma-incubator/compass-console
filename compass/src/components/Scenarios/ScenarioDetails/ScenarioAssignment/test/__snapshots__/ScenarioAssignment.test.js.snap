// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScenarioAssignment Renders with minimal props 1`] = `
<GenericListMocked
  actions={
    Array [
      Object {
        "handler": [Function],
        "name": "Delete",
      },
    ]
  }
  entries={
    Array [
      Object {
        "scenarioName": "test-scenario",
        "selector": Object {
          "key": "test-key",
          "value": "test-value",
        },
      },
    ]
  }
  headerRenderer={[Function]}
  notFoundMessage="No Automatic Scenario Assignment for this Scenario"
  rowRenderer={[Function]}
  showSearchField={false}
  title="Automatic Scenario Assignment"
/>
`;
